Issue28459 dvfs doc (#17061)
nrfconnect/sdk-nrf: applications/nrf_desktop/doc/dvfs.rst

.. _nrf_desktop_dvfs:

DVFS module
###########

.. contents::
   :local:
   :depth: 2

> The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module is responsible for switching the frequency and voltage according to the application needs.

Reviewer: Will the Arch diagram also require an update? - https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/applications/nrf_desktop/description.html#module_and_component_overview

Author: We have a ticket for that: https://nordicsemi.atlassian.net/browse/NCSDK-28811

> The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module is responsible for switching the frequency and voltage according to the application needs.

Reviewer 2: Can we mention that the module actually controls clock frequency (voltage changes are done under the hood by the used clock driver)?

DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

> In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when the application performs less computation-challenging activities such as Bluetooth LE advertising.

Reviewer 3: +maybe `Bluetooth® LE` , since it's the first occurrence

> In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when the application performs less computation-challenging activities such as Bluetooth LE advertising.

Reviewer 2: Can we rephrase it a bit to something like:
`By default, system uses the lowest available frequency to save power. If application performs computation-consuming tasks, the DVFS module will increase frequency for better performance.`?

That would be more in line with the implementation

Author: It is not true what you are writing: "By default we use 320MHz so that's why I would state that we lower frequency in some cases and not the other way around."

Reviewer 2: I would focus on the use-case where the module is enabled and actively controls the frequency here.
Maybe we could rephrase a bit to prevent misunderstandings: `By default, the module sets the lowest available frequency to save power. ...`

Reviewer 2: You can still mention that disabling the module leads to using `320MHz` frequency all the time (IMHO this is out of scope of this doc file)

Author: > I would focus on the use-case where the module is enabled and actively controls the frequency here. Maybe we could rephrase a bit to prevent misunderstandings: `By default, the module sets the lowest available frequency to save power. ...`

It is thea same as this sentence: "In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when the application performs less computation-challenging activities such as Bluetooth LE advertising."

It is just more generic

Reviewer 2: I would not insist here, we can keep it as is. I just tried to make this part more in line with the implementation.

> However, if performing more computation-consuming tasks nRF Desktop DVFS module will dynamically change to a higher frequency to increase performance.

Reviewer 3: Additional suggestion - maybe something along these lines for better flow (incorporates Pekka's suggestion)?
```suggestion
However, for more compute-intensive tasks, the nRF Desktop DVFS module will dynamically switch to a higher frequency to enhance performance.
```

> Requesting for frequency change is done by requesting clock from clock control module.

Reviewer 2: ```
The application module requests clock frequency change using API of the clock control driver.
The driver submits a request to the sysctrl core through the nrfs DVFS service.
The core controls both clock frequency and voltage.
```

> Requesting for frequency change is done by requesting clock from clock control module.

Author: Already changed after Mareks remark

> Clock control module is responsible for choosing the right clock frequency and voltage.

Author: Already changed after Mareks remark

> The DVFS is set by the sysctrl core and Clock control module requests are send through nrfs service DVFS.

Author: Already changed after Mareks remark

.. note::

>    In the current solution, it is assumed that no other module requests frequency changes.

Author: We should leave version with assume as we do not control other modules

>    The nRF Desktop DVFS module is the sole entity responsible for managing frequency and voltage adjustments.

Reviewer 2: Can we also mention that:
`Controlling clock from other sources would result in a module fatal error.`

Supported devices
*****************

> The module is only available for nrf54h20 SoC.

Reviewer 4: ```suggestion
The module is only available for the nRF54H20 SoC.
```

> Each device has a different set of available operating points.

Reviewer 2: Can we skip this sentence as we have support only for one SoC with DVFS?

> For example the nrf54h20 SoC has the following available operating points (in documentation called frequency):

Reviewer 2: `The nRF54H20 SoC ...`

> 1. DVFS_FREQ_HIGH - high frequency - 320 MHz, 0.8 V
> #. DVFS_FREQ_MEDLOW - medium-low frequency - 128 MHz, 0.6V
> #. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Reviewer 4: ```suggestion
* DVFS_FREQ_HIGH - high frequency - 320 MHz, 0.8 V
* DVFS_FREQ_MEDLOW - medium-low frequency - 128 MHz, 0.6V
* DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V
```

Reviewer 3: Should we capitalize the text after dash?

> #. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Reviewer 2: Can we mention that the application module might request any clock frequency, but it would be rounded down to one of the supported values by the driver? (that's why we define a set of supported frequencies in scope of the application module)

Module events
*************

.. include:: event_propagation.rst
    :start-after: table_dvfs_start
    :end-before: table_dvfs_end

.. note::
    |nrf_desktop_module_event_note|

Configuration
*************

> The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module is enabled by default on supported board targets and can be disabled by setting the :option:`CONFIG_DESKTOP_DVFS` Kconfig option to ``n``.

Reviewer 4: ```suggestion
The nRF Desktop DVFS module is enabled by default on supported board targets.
To disable it, set the :option:`CONFIG_DESKTOP_DVFS` Kconfig option to ``n``.
```

> The module can be configured using the following Kconfig options:

Reviewer 2: This sentence could be removed

> The nRF Desktop DVFS module defines independent states that have different requirements for the DVFS operating point.

Reviewer 2: ```suggestion
The nRF Desktop DVFS module defines independent application states that require higher clock frequency.
```

Author: My version seams better IMO.

Reviewer 2: Can we add just the `application states` part then?

Reviewer 2: I tried to separate the module from voltage control (we started using the clock control APIs; voltage is controlled under the hood from module's perspective), but I noticed that you prefer to refer to `DVFS operating points` etc. I would not insist here.

Author: To be precise frequency is also controlled under the hood. We have points(freq, V), which are ordered - the way we call them doesn't matter. We could name them a,b,c,...   and let user select one. The way you try to present this shows false impression that user have possibility to choose any frequency. We call them frequency because for most people it is the most important part.

Author: > Can we add just the `application states` part then?

isn't it already done?

Reviewer 2: > To be precise frequency is also controlled under the hood. We have points(freq, V), which are ordered - the way we call them doesn't matter. We could name them a,b,c,... and let user select one. The way you try to present this shows false impression that user have possibility to choose any frequency. We call them frequency because for most people it is the most important part.

I am looking from clock control API user perspective (application module is clock control API user). From this perspective we control the clock frequency. Keep in mind that even back in time when we used enum, the enum names were related to frequency (`DVFS_FREQ_xyz`). So we control frequency here and voltage is scaled under the hood (application module never explicitly selected an oppoint).

> The module listens for Even Manager events that are sent and turns state on/off according to the event.

Reviewer 3: Event* Manager

> The module listens for Even Manager events that are sent and turns state on/off according to the event.

Reviewer 2: `Event Manager`

It support the following DVFS states:

> * ``DVFS_STATE_INITIALIZING`` - State indicating that core initialization is in process.

Reviewer 2: `... application initialization is in progress.`

* ``DVFS_STATE_LLPM_CONNECTED`` - State indicating that Bluetooth Low Latency Packet Mode is active.
* ``DVFS_STATE_USB_CONNECTED`` - State indicating that USB is connected.
* ``DVFS_STATE_CONFIG_CHANNEL`` - State indicating that Config Channel is active.
* ``DVFS_STATE_SMP_TRANSFER`` - State indicating that SMP transfer is active.

Each DVFS state can be configured using the following Kconfig options:

> * ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ENABLE`` - Enable DVFS state <STATE>.
> * ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ`` - Choice of active frequency of state <STATE>.

Reviewer 3: ```suggestion
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ENABLE`` - Enables DVFS state <STATE>.
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ`` - Sets the active frequency for state <STATE>.
```

Reviewer 3: Just wondering, but is <STATE> how this should be indicated? Should we express it graphically in some way, e.g., ``<STATE>``? What do you think @peknis ?

Reviewer 2: Can we mention that the choice defines frequency `requested` by the state? Or minimal frequency required when the state is active?

    The possible values are:  ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_HIGH`` and ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_MEDLOW``.

> * ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Timeout in milliseconds specifying time from receiving on event to moment the state <STATE> stops requesting its active frequency.

Reviewer 4: ```suggestion
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Timeout in milliseconds, specifying the time from receiving an event to the moment the state <STATE> stops requesting its active frequency.
```

Reviewer 3: to align with other 2 bullets. Also, incorporates Pekka's suggestion
```suggestion
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Specifies the timeout in milliseconds, defining the duration from receiving an event to the moment the state <STATE> stops requesting its active frequency.
```

>     When the timeout is reached, the state will go back to a nonactive state potentially enabling the system to go to lower frequency.

Reviewer 3: non-active* for better readability

>     If timeout is set to 0 the <STATE> will not cease requesting its active frequency until receiving off event.

Reviewer 3: Maybe: 
```suggestion
    If the timeout is set to ``0`` the <STATE> will continue requesting its active frequency until it receives an off event.
```
includes Pekka's suggestion

Each change request is retried if it fails due to known reason.
The module will retry the request up to the number of times specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_COUNT`.
Error counter is only zeroed when preforming correct DVFS change request.

> Failing frequency change requests can be caused by another frequency change request being in progress on sysctrl - error ``-EBUSY`` or by the nrfs DVFS service being not yet initialized - error ``-EAGAIN``.

Reviewer 3: Maybe we could add "resulting in" before the errors are mentioned?

> If the frequency change request is in progress, the module will retry the request after a timeout specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS`.

Reviewer 2: The following syntax needs to be used for nRF Desktop specific Kconfig options:
```
:ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>`
```

If the nrfs service DVFS is not yet initialized, the nRF Desktop DVFS module will retry the request after a timeout specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS`.
If the DVFS frequency request returns an error different than ``-EBUSY`` or ``-EAGAIN``, the module will report ``MODULE_STATE_ERROR`` and set ``module_state`` to ``STATE_ERROR``.
nRF Desktop DVFS module will also report MODULE_STATE_ERROR and set ``module_state`` to ``STATE_ERROR`` if the number of retries exceeds the value specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_COUNT`.
In ``module_state`` in ``STATE_ERROR`` the module will not change the frequency and voltage.

> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.

Reviewer 3: ```suggestion
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Specifies the timeout in milliseconds, after which the nRF Desktop DVFS module will retry DVFS frequency change.
```
to align

Author: Already removed that part from doc

> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.

Reviewer 2: The timeout values are already described in section above (where you describe the handled error codes). No need to describe them again here.

> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.
>    This timeout is applied in case another DVFS change request is still in progress which causes the current request to fail.

Reviewer 4: ```suggestion
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying the time after which the nRF Desktop DVFS module will retry a DVFS frequency change.
   This timeout is applied if another DVFS change request is still in progress, which causes the current request to fail.
```

   The default value is 1 ms.
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.
   This timeout is applied in case DVFS is not yet initialized which causes the current request to fail.
   The default value is 500 ms.

> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_COUNT <config_desktop_app_options>` - Number of retries of DVFS frequency change after which nRF Desktop DVFS module will report MODULE_STATE_ERROR.

Reviewer 4: ```suggestion
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_COUNT <config_desktop_app_options>` - Number of retries of DVFS frequency change after which the nRF Desktop DVFS module will report ``MODULE_STATE_ERROR``.
```

>    The default value is 5.

Reviewer: ```suggestion
   The default value is ``5``.
```

> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.
>    This timeout is applied in case another DVFS change request is still in progress which causes the current request to fail.
>    The default value is 1 ms.
> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.
>    This timeout is applied in case DVFS is not yet initialized which causes the current request to fail.
>    The default value is 500 ms.
> * :ref:`CONFIG_DESKTOP_DVFS_RETRY_COUNT <config_desktop_app_options>` - Number of retries of DVFS frequency change after which nRF Desktop DVFS module will report MODULE_STATE_ERROR.
>    The default value is 5.

Reviewer: ```suggestion
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.
  This timeout is applied in case another DVFS change request is still in progress which causes the current request to fail.
  The default value is 1 ms.
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS <config_desktop_app_options>` - Timeout in milliseconds specifying time after which nRF Desktop DVFS module will retry DVFS frequency change.
  This timeout is applied in case DVFS is not yet initialized which causes the current request to fail.
  The default value is 500 ms.
* :ref:`CONFIG_DESKTOP_DVFS_RETRY_COUNT <config_desktop_app_options>` - Number of retries of DVFS frequency change after which nRF Desktop DVFS module will report MODULE_STATE_ERROR.
  The default value is 5.
```
This comment only fixes the alignment issue. Please address Pekka's and Anna's comment on these lines.

Implementation details
**********************

> The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module has two state tracing entities.

Reviewer 4: ```suggestion
The nRF Desktop DVFS module has the following two state tracing entities:
```

> The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module has two state tracing entities.

Reviewer 2: Not sure if we should actually describe that in detail (it's quite straightforward when you look into code, other app modules do the same + it increases maintenance for us)

Author: I would leave it to make clear that we have two state entities so people can not mistaken them.

First ``module_state`` is used to track the state of the nRF Desktop DVFS module.
It indicates if the module is ready or if an error has occurred.

> Second ``dfvs_requests_state_bitmask`` is used to track separately the state of each DVFS state.

Reviewer 2: `... to track application DVFS states`.

Each DVFS state is represented by a bit in the bitmask.
If the bit is set, the state is active.

> The module listens for events that are sent and applies state activity according to the event.

Reviewer 2: `The module listens for Event Manager events related to a given state and updates state accordingly`.

> The module then checks which states are active and sets the highest frequency of all active states.

Reviewer 2: `and sets the highest frequency that is required by at least one state`.

.. _nrf_desktop_dvfs:

DVFS module
###########

.. contents::
   :local:
   :depth: 2

The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module is responsible for switching the frequency and voltage according to the application needs.

> The module chooses the highest frequency required by the application states and selects voltage accordingly.

Author: To be precise frequency is also controlled under the hood. We have points(freq, V), which are ordered - the way we call them doesn't matter. We could name them a,b,c,... and let user select one. The way you try to present this shows false impression that user have possibility to choose any frequency. We call them frequency because for most people it is the most important part.

> The module chooses the highest frequency required by the application states and selects voltage accordingly.

Reviewer 2: You set frequency through the API (passing it explicitly as a parameter). The frequency rounding and voltage update is done under the hood by the underlying drivers' implementations.

DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when the application performs less computation-challenging activities such as Bluetooth LE advertising.
However, if performing more computation-consuming tasks nRF Desktop DVFS module will dynamically change to a higher frequency to increase performance.

The nRF Desktop DVFS module requests clock frequency change using API of the clock control driver.

> The driver is responsible for choosing the optimal clock frequency and voltage.

Reviewer 2: The driver actually seems not to know anything about the voltage. It's more like: the driver controls frequency and sysctrl core aligns voltage when setting the frequency.

Author: That's no entirely true but I don't have time to argue. Let it be your way, if you must

Reviewer 2: I am not sure about responsibility split between nrfs service on app core and code running on sysctrl core (so I might be inaccurate here). Still, if we inform what happens under the hood, we should not provide false impression (the clock control driver really does not choose voltage :) ).

The driver submits a request to the sysctrl core through the nrfs DVFS service.
The sysctrl core controls both clock frequency and voltage.

.. note::
   In the current solution, it is assumed that no other module requests frequency changes.
   The nRF Desktop DVFS module is the sole entity responsible for managing frequency and voltage adjustments.
   Controlling clock from other sources would result in a module fatal error.

Supported devices
*****************

The module is only available for nRF54H20 SoC.

> For example the nRF54H20 SoC has the following available operating points (in documentation called frequency):

Reviewer 2: Can we remove the `For example ...` ? (as suggested in: https://github.com/nrfconnect/sdk-nrf/pull/17061#discussion_r1820204039). Previous line already mentions that only a single SoC is supported by the module.

  The possible values are:  ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_HIGH`` and ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_MEDLOW``.
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Timeout in milliseconds specifying time from receiving on event to moment the state <STATE> stops requesting its active frequency.
  When the timeout is reached, the state will go back to a nonactive state potentially enabling the system to go to lower frequency.
  If timeout is set to 0 the <STATE> will not cease requesting its active frequency until receiving off event.
  Some of the tracked application states have associated events which inform when the state is turned on and off (e.g. USB connection).
  Other states have associated application events emitted periodically when active.
  There is no event explicitly informing that state is no longer active (e.g. config channel).
  For such states, we need to define a timeout that is used to turn off the tracked state after associated application events are no longer emitted.

Frequency change retry
======================

Each frequency change request is retried if it fails due to known reason.

> Failing frequency change requests can be caused by:

Reviewer 2: Need empty line before the bullet list

* Another frequency change request being in progress on sysctrl - error ``-EBUSY``.
  In that case, the module will retry the request after a timeout specified by :ref:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS <config_desktop_app_options>`.
  The default timeout value is 1 ms.
* The nrfs DVFS service being not yet initialized - error ``-EAGAIN``.
  In that case, the nRF Desktop DVFS module will retry the request after a timeout specified by :ref:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS <config_desktop_app_options>`.
  The default timeout value is 500 ms.
* Other errors are not retried.
  If the DVFS frequency request returns an error different than ``-EBUSY`` or ``-EAGAIN``, the module will report :c:enum:`MODULE_STATE_ERROR` and set ``module_state`` to ``STATE_ERROR``.

The module will retry the request up to the number of times specified by :ref:`CONFIG_DESKTOP_DVFS_RETRY_COUNT <config_desktop_app_options>`.
The default value of retries is 5.
Error counter is only zeroed when preforming correct DVFS change request.
nRF Desktop DVFS module will also report :c:enum:`MODULE_STATE_ERROR` and set ``module_state`` to ``STATE_ERROR`` if the number of retries is exceeded.

> In ``module_state`` in ``STATE_ERROR`` the module will not change the frequency and voltage.

Reviewer 2: ```suggestion
In ``STATE_ERROR`` the module will not change the frequency and voltage.
```

:c:enum:`MODULE_STATE_ERROR` reported by the nRF Desktop DVFS module would lead to fatal application error.

Implementation details
**********************

The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module has two state tracing entities.

First ``module_state`` is used to track the state of the nRF Desktop DVFS module.
It indicates if the module is ready or if an error has occurred.

Second ``dfvs_requests_state_bitmask`` is used to track separately the state of each application DVFS state.
Each DVFS state is represented by a bit in the bitmask.
If the bit is set, the state is active.
The module listens for Event Manager events related to a given state and updates state accordingly.
The module then checks which states are active and sets the highest frequency that is required by at least one state.

> If no state is active, the module sets the lowest frequency: ``DVFS_FREQ_LOW``.

Reviewer 2: `CONFIG_DESKTOP_DVFS_FREQ_LOW`?

However, if performing more computation-consuming tasks nRF Desktop DVFS module will dynamically change to a higher frequency to increase performance.

The nRF Desktop DVFS module requests clock frequency change using API of the clock control driver.
The driver is responsible for choosing the optimal clock frequency.
The driver submits a request to the sysctrl core through the nrfs DVFS service.
The sysctrl core controls both clock frequency and voltage and chooses corresponding voltage to requested frequency.

.. note::
   In the current solution, it is assumed that no other module requests frequency changes.
   The nRF Desktop DVFS module is the sole entity responsible for managing frequency and voltage adjustments.
   Controlling clock from other sources would result in a module fatal error.

Supported devices
*****************

The module is only available for nRF54H20 SoC.

The nRF54H20 SoC has the following available operating points (in documentation called frequency):

1. DVFS_FREQ_HIGH - high frequency - 320 MHz, 0.8 V
#. DVFS_FREQ_MEDLOW - medium-low frequency - 128 MHz, 0.6V

> #. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Reviewer 5: We do not use the DVFS API directly anymore, maybe we should just specify it as HIGH/MEDLOW/LOW as the enum does not matter here (it is used under the hood)?

Author: It matters in a way that you cannot have different values. I would leave it as it is to show where from those values came from.

The default value of retries is 5.
Error counter is only zeroed when preforming correct DVFS change request.
nRF Desktop DVFS module will also report :c:enum:`MODULE_STATE_ERROR` and set ``module_state`` to ``STATE_ERROR`` if the number of retries is exceeded.

In ``STATE_ERROR`` the module will not change the frequency and voltage.
:c:enum:`MODULE_STATE_ERROR` reported by the nRF Desktop DVFS module would lead to fatal application error.

Implementation details
**********************

The nRF Desktop Dynamic Voltage and Frequency Scaling (DVFS) module has two state tracing entities.

First ``module_state`` is used to track the state of the nRF Desktop DVFS module.
It indicates if the module is ready or if an error has occurred.

Second ``dfvs_requests_state_bitmask`` is used to track separately the state of each application DVFS state.
Each DVFS state is represented by a bit in the bitmask.
If the bit is set, the state is active.
The module listens for Event Manager events related to a given state and updates state accordingly.
The module then checks which states are active and sets the highest frequency that is required by at least one state.

> If no state is active, the module sets the lowest frequency: ``CONFIG_DESKTOP_DVFS_FREQ_LOW``.

Reviewer 2: Please fix ref to an app-specific Kconfig option.

.. note::
   In the current solution, it is assumed that no other module requests frequency changes.
   The nRF Desktop DVFS module is the sole entity responsible for managing frequency and voltage adjustments.
   Controlling clock from other sources would result in a module fatal error.

Supported devices
*****************

The module is only available for the nRF54H20 SoC.

The nRF54H20 SoC has the following available operating points (called frequency in the documentation):

* DVFS_FREQ_HIGH - High frequency - 320 MHz, 0.8 V
* DVFS_FREQ_MEDLOW - Medium-low frequency - 128 MHz, 0.6V
* DVFS_FREQ_LOW - Low frequency - 64 MHz, 0.5V

The nRF Desktop DVFS module might request any clock frequency, but it would be rounded up by the driver to one of the values from available operating points.

Module events
*************

.. include:: event_propagation.rst
    :start-after: table_dvfs_start
    :end-before: table_dvfs_end

.. note::
    |nrf_desktop_module_event_note|

Configuration
*************

The nRF Desktop DVFS module is enabled by default on supported board targets.

> To disable it, set the :option:`CONFIG_DESKTOP_DVFS` Kconfig option to ``n``.

Reviewer 2: Wrong Kconfig option ref syntax

Tracked application states
==========================

> The nRF Desktop DVFS module defines independent application states that have different requirements for each DVFS operating point.

Reviewer 2: `have different requirements in terms of DVFS operating point`? `that require using different DVFS operating point`?

Reviewer 2: Also I would use term `frequency` instead of `DVFS operating point` (we have a paragraph that explains how `frequency` maps to `DVFS operating point` and what happens under the hood - it should be enough). Still, I would not insist

* ``DVFS_STATE_USB_CONNECTED`` - USB is connected
* ``DVFS_STATE_CONFIG_CHANNEL`` - :ref:`nrf_desktop_config_channel` is active
* ``DVFS_STATE_SMP_TRANSFER`` - DFU image transfer over BLE SMP is active
  For details about SMP DFU integration in nRF Desktop, see :ref:`nrf_desktop_dfu_mcumgr`.

You can configure each DVFS state using the following Kconfig options:

* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ENABLE`` - Enables DVFS state <STATE> handling.
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ`` - Choice of minimal active frequency requested by state <STATE>.
  The possible values are ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_HIGH`` and ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_MEDLOW``.
* ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Specifies the timeout in milliseconds, defining the duration from receiving an on event to the moment the state <STATE> stops requesting its active frequency.
  When the timeout is reached, the state returns to a non-active state potentially enabling the system to use lower frequency.
  If the timeout is set to ``0`` the <STATE> will continue requesting its active frequency until it receives an off event.

>   Some of the tracked application states have associated events which inform when the state is turned on and off (e.g. USB connection).

Reviewer: ```suggestion
  Some of the tracked application states have associated events which inform when the state is turned on and off (for example, USB connection).
```

Reviewer 4: ... events that (instead of which)

  Other states have associated application events emitted periodically when active.

>   There is no event explicitly informing that state is no longer active (e.g. config channel).

Reviewer: ```suggestion
  There is no event explicitly informing that state is no longer active (for example, config channel).
```

.. _nrf_desktop_dvfs:

DVFS module
###########

.. contents::
   :local:
   :depth: 2

The Dynamic Voltage and Frequency Scaling (DVFS) module is responsible for switching the frequency and voltage according to the application needs.

> It is a way to optimize power consumption by adjusting the frequency and voltage of the core according to the current system state.

Reviewer 5: I think this sentence might not be necessary and is a bit redundant (half of the info is in the sentence before - what the DFVS does, half in the sentence below - the reason why we use it so power consumption)

DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

> In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when connected using classic BLE as it is enough to run this mode.

Author: When not connected we should go to sleep and do not have any frequency.

Maybe just removing "connected", as the process of advertising is also classic BLE?

However, if performing more computation-consuming tasks DVFS module will dynamically change to a higher frequency to increase performance.

Available operating points (in documentation called frequency):

1. DVFS_FREQ_HIGH - high frequency - 320 MHz, 0.8 V
#. DVFS_FREQ_MEDLOW - medium-low frequency - 128 MHz, 0.6V

> #. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Author: IMO placing values is important - leaving just names like FREQ_MEDLO gives little information.
Currently we have only nrf54h20. Lets deal with it when we will have another SoC with DVFS support - then maybe we can create some table.

> #. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Author: I was thinking about it and I guess we will have to rework it heavily due to change in api when we move to clock api.
Till then I would leave it as it is.

> #. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Reviewer 2: I see there seem to be no bigger doc section we could link to currently. Still, please at least mention that voltages and frequencies are specific to nRF54H20 (or skip mentioning these values completely).

Module events
*************

.. include:: event_propagation.rst
    :start-after: table_dvfs_start
    :end-before: table_dvfs_end

.. note::
    |nrf_desktop_module_event_note|

Configuration
*************

> The Dynamic Voltage and Frequency Scaling (DVFS) module is enabled by default for nrf54h20 Soc and can be disabled by setting the :option:`CONFIG_DESKTOP_DVFS` Kconfig option to ``n``.
> The module is only available for nrf54h20 SoC.

Reviewer 5: consider changing the order of info (IMO it would be better to firstly now if it is even supported on specific platform rather than how to configure it)
```suggestion
The Dynamic Voltage and Frequency Scaling (DVFS) module is available only for the nRF54H20 SoC.
It is enabled by default and can be disabled using the :option:`CONFIG_DESKTOP_DVFS` Kconfig option.
```

Author: I moved "The module is only available for nrf54h20 SoC."
before list of available frequencies so will leave first sentence as it is here.

The DVFS module can be configured using the following Kconfig options:

1. :option:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS` - Timeout in milliseconds specifying time after which DVFS module will retry DVFS frequency change.
   This timeout is applied in case another DVFS change request is still in progress which causes the current request to fail.
   The default value is 1 ms.
#. :option:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS` - Timeout in milliseconds specifying time after which DVFS module will retry DVFS frequency change.
   This timeout is applied in case DVFS is not yet initialized which causes the current request to fail.
   The default value is 500 ms.
#. :option:`CONFIG_DESKTOP_DVFS_RETRY_COUNT` - Number of retries of DVFS frequency change after which DVFS module will report MODULE_STATE_ERROR.
   The default value is 5.

Each DVFS state can be configured using the following Kconfig options:

> 1. `CONFIG_DESKTOP_DVFS_STATE_<STATE>_ENABLE` - Enable DVFS state <STATE>.

Reviewer 5: should it be double backtick? same in other options
```suggestion
1. ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ENABLE`` - Enable DVFS state <STATE>.
```

#. `CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS` - Timeout in milliseconds specifying time after which the DVFS module returns to low frequency after entering <STATE>.
#. `CONFIG_DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ` - Default active frequency of <STATE>.

Currently supported DVFS states are:

> 1. DVFS_STATE_INITIALIZING - State indicating that core initialization is in process.

Reviewer 5: maybe put each state in in double backtick also?

#. DVFS_STATE_LLPM_CONNECTED - State indicating that Low Latency Packet Mode is active.
#. DVFS_STATE_USB_CONNECTED - State indicating that USB is connected.
#. DVFS_STATE_CONFIG_CHANNEL - State indicating that Config Channel is active.
#. DVFS_STATE_SMP_TRANSFER - State indicating that SMP transfer is active.

Implementation details
**********************

The Dynamic Voltage and Frequency Scaling (DVFS) has two state tracing entities.

> First `module_state` is used to track the state of the DVFS module.

Reviewer 5: I think that all variables/code-related names should be in double backticks.

> It states if the DVFS module is ready or did it encounters an error.

Reviewer 5: ```suggestion
It indicates if the DVFS module is ready or if an error has occurred.
```

Second `dfvs_requests_state_bitmask` is used to track separately the state of each DVFS state.
Each DVFS state is represented by a bit in the bitmask.
If the bit is set, the state is active.
DVFS module listens for events that are sent and applies state activity according to the event.
DVFS module then checks which states are active and sets the highest frequency of all active states.

> If no state is active, the module sets the lowest frequency: DVFS_FREQ_LOW.

Reviewer 5: you have already described all possible frequency states, consider:
```suggestion
If no state is active, the module sets the lowest frequency.
```

Author: I insist on being specific here. Drooping this information can be misleading as one might wonder if I mean Low freq or there is some option where clock is turned off etc.

Each change request is retried if it fails.
The module will retry the request up to the number of times specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_COUNT`.
Failing frequency change request can be caused by another frequency change request in progress or by the DVFS module being not yet initialized.
If the frequency change request is in progress, the module will retry the request after a timeout specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS`.
If the DVFS module is not yet initialized, the module will retry the request after a timeout specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS`.

> If the DVFS frequency request returns an error different than `EBUSY` or `EAGAIN`, the module will report MODULE_STATE_ERROR and set `module_state` to `STATE_ERROR`.

Reviewer 5: shouldn't it be
```
... ``-EBUSY`` or ``-EAGAIN`` ...
```
?

Also maybe link those error codes with the sentences that describes the behaviour of the change request function above?

.. _nrf_desktop_dvfs:

DVFS module
###########

.. contents::
   :local:
   :depth: 2

The Dynamic Voltage and Frequency Scaling (DVFS) module is responsible for switching the frequency and voltage according to the application needs.

> DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

Author: For now we don't have such documentation. The best we can find is https://docs.zephyrproject.org/apidoc/latest/group__clock__control__interface.html#details

But it is not enough to link it here IMO. Lets create a follow-up ticket to add it in future.

> DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

Author: https://nordicsemi.atlassian.net/browse/NCSDK-29951

> DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

Reviewer 2: If impact is not high, we should desist from using `critical` word here I think

> DVFS is critical for balancing performance and power consumption, especially in battery-operated devices.

Reviewer 5: In some way each one of the application modules is critical :D I would desist from using this word and just inform like Marek suggested

> In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when using classic BLE as it is enough to run this mode.

Reviewer 2: You can even state that we use low frequency when no computation-challenging operation is performed (to save power). I would suggest to provide an example of activity that actually requires higher performance instead.

> In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when using classic BLE as it is enough to run this mode.

Reviewer 2: E.g. you can refer to achieving 8000 Hz report rate over USB.

> In the nRF Desktop, the DVFS module enables the system to use the lowest possible frequency and voltage when using classic BLE as it is enough to run this mode.

Author: By default we use 320MHz so that's why I would state that we lower frequency in some cases and not the other way around.

However, if performing more computation-consuming tasks DVFS module will dynamically change to a higher frequency to increase performance.

> The module is only available for nrf54h20 SoC.

Reviewer 2: I would suggest to simply state that DVFS feature must be supported by the SoC and add nRF54H20 as an example of SoC that supports it (I prefer to avoid making sections and subsections related to SoCs/boards to avoid maintaining them later). We use no SoC-specific code here too (we just rely on fact that DVFS API is available).

Available operating points (in documentation called frequency):

1. DVFS_FREQ_HIGH - high frequency - 320 MHz, 0.8 V
#. DVFS_FREQ_MEDLOW - medium-low frequency - 128 MHz, 0.6V
#. DVFS_FREQ_LOW - low frequency - 64 MHz, 0.5V

Module events
*************

.. include:: event_propagation.rst
    :start-after: table_dvfs_start
    :end-before: table_dvfs_end

.. note::
    |nrf_desktop_module_event_note|

Configuration
*************

> The Dynamic Voltage and Frequency Scaling (DVFS) module is enabled by default for nrf54h20 Soc and can be disabled by setting the :option:`CONFIG_DESKTOP_DVFS` Kconfig option to ``n``.

Reviewer 2: Or maybe `on board targets where SoC supports DVFS feature in hardware`?

> The Dynamic Voltage and Frequency Scaling (DVFS) module is enabled by default for nrf54h20 Soc and can be disabled by setting the :option:`CONFIG_DESKTOP_DVFS` Kconfig option to ``n``.

Author: IMO on supported board targets is precise enough.

> The DVFS module can be configured using the following Kconfig options:

Reviewer 2: Can we introduce a description on how module works before going into details related to configuration? That would make config options documentation far easier to understand.

1. :option:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS` - Timeout in milliseconds specifying time after which DVFS module will retry DVFS frequency change.
   This timeout is applied in case another DVFS change request is still in progress which causes the current request to fail.
   The default value is 1 ms.
#. :option:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS` - Timeout in milliseconds specifying time after which DVFS module will retry DVFS frequency change.
   This timeout is applied in case DVFS is not yet initialized which causes the current request to fail.
   The default value is 500 ms.
#. :option:`CONFIG_DESKTOP_DVFS_RETRY_COUNT` - Number of retries of DVFS frequency change after which DVFS module will report MODULE_STATE_ERROR.
   The default value is 5.

Each DVFS state can be configured using the following Kconfig options:

1. ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ENABLE`` - Enable DVFS state <STATE>.

> #. ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Timeout in milliseconds specifying time after which the DVFS module returns to low frequency after entering <STATE>.

Reviewer 6: Another improvement is to divide the states into two groups: the ones that timeout-based and the ones that are driven by on/off events

> #. ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_TIMEOUT_MS`` - Timeout in milliseconds specifying time after which the DVFS module returns to low frequency after entering <STATE>.

Author: We do not disable or divide states. It is done in other way - every state has this functionality just not all states use it it is configurable - see "If timeout is set to 0 the <STATE> will not cease requesting its active frequency until receiving off event."

#. ``CONFIG_DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ`` - Choice of active frequency of <STATE>.
    The possible values are:  ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_HIGH`` and ``DESKTOP_DVFS_STATE_<STATE>_ACTIVE_FREQ_MEDLOW``.

Currently supported DVFS states are:

1. ``DVFS_STATE_INITIALIZING`` - State indicating that core initialization is in process.
#. ``DVFS_STATE_LLPM_CONNECTED`` - State indicating that Low Latency Packet Mode is active.
#. ``DVFS_STATE_USB_CONNECTED`` - State indicating that USB is connected.
#. ``DVFS_STATE_CONFIG_CHANNEL`` - State indicating that Config Channel is active.
#. ``DVFS_STATE_SMP_TRANSFER`` - State indicating that SMP transfer is active.

Implementation details
**********************

> The Dynamic Voltage and Frequency Scaling (DVFS) has two state tracing entities.

Reviewer 6: ```suggestion
The Dynamic Voltage and Frequency Scaling (DVFS) module has two state tracing entities.
```

> The Dynamic Voltage and Frequency Scaling (DVFS) has two state tracing entities.

Reviewer 6: We have three different entities here and I would be careful with wording:

- the DVFS module (nRF Desktop module)
- the DVFS capability (HW feature)
- the DVFS service (part of nrfs to interact with the HW feature via SCFW)

Author: Hope I clarified it:
the DVFS module (nRF Desktop module) -  "nRF Desktop DVFS "
the DVFS capability (HW feature)  - "DVFS"
the DVFS service (part of nrfs to interact with the HW feature via SCFW) - "nrfs service DVFS"

First ``module_state`` is used to track the state of the DVFS module.
It indicates if the DVFS module is ready or if an error has occurred.

Second ``dfvs_requests_state_bitmask`` is used to track separately the state of each DVFS state.
Each DVFS state is represented by a bit in the bitmask.
If the bit is set, the state is active.
DVFS module listens for events that are sent and applies state activity according to the event.
DVFS module then checks which states are active and sets the highest frequency of all active states.

> If no state is active, the module sets the lowest frequency: ``DVFS_FREQ_LOW``.

Reviewer 6: Please mention the SW entity (nrfs and DVFS service) that is used to request the actual frequency change. Also, add a ref link to it

> Each change request is retried if it fails.

Reviewer 2: Maybe we could move part of this to configuration section to make config options' doc easier to understand.

Reviewer 2: This section could be left only with information that are small details related to implementation

The module will retry the request up to the number of times specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_COUNT`.

> Failing frequency change request can be caused by another frequency change request in progress error ``-EBUSY`` or by the DVFS module being not yet initialized error ``-EAGAIN``.

Reviewer 6: ```suggestion
Failing frequency change requests can be caused by another frequency change request in progress error ``-EBUSY`` or by the DVFS service being not yet initialized error ``-EAGAIN``.
```

If the frequency change request is in progress, the module will retry the request after a timeout specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_BUSY_TIMEOUT_MS`.

> If the DVFS module is not yet initialized, the module will retry the request after a timeout specified by :option:`CONFIG_DESKTOP_DVFS_RETRY_INIT_TIMEOUT_MS`.

Reviewer 6: This is the DVFS module, you use the same name to refer to two different things.
